bulldozer/merge.go

```diff
@@ -89,78 +117,75 @@ func MergePR(ctx context.Context, pullCtx pull.Context, client *github.Client, m
 		for i := 0; i < MaxPullRequestPollCount; i++ {
 			<-ticker.C
 
-			pr, _, err := client.PullRequests.Get(ctx, pullCtx.Owner(), pullCtx.Repo(), pullCtx.Number())
+			mergeState, err := pullCtx.MergeState(ctx)
 			if err != nil {
-				logger.Error().Err(errors.WithStack(err)).Msgf("Failed to retrieve pull request %q", pullCtx.Locator())
+				logger.Error().Err(err).Msgf("Failed to get merge state for %q", pullCtx.Locator())
 				return
 			}
 
-			if pr.GetState() == "closed" {
+			if mergeState.Closed {
 				logger.Debug().Msg("Pull request already closed")
 				return
 			}
 
-			if pr.Mergeable == nil {
+			if mergeState.Mergeable == nil {
 				logger.Debug().Msg("Pull request mergeability not yet known")
 				continue
 			}
 
-			if !pr.GetMergeable() {
+			if !*mergeState.Mergeable {
 				logger.Debug().Msg("Pull request is not mergeable")
 				return
 			}
 
 			// Try a merge, a 405 is expected if required reviews are not satisfied
 			logger.Info().Msgf("Attempting to merge pull request with method %s", mergeOpts.MergeMethod)
-			result, _, err := client.PullRequests.Merge(ctx, pullCtx.Owner(), pullCtx.Repo(), pullCtx.Number(), commitMessage, mergeOpts)
+			sha, err := merger.Merge(ctx, pullCtx, commitMessage, mergeOpts)
 			if err != nil {
-				gerr, ok := err.(*github.ErrorResponse)
+				gerr, ok := errors.Cause(err).(*github.ErrorResponse)
 				if !ok {
-					logger.Error().Err(errors.WithStack(err)).Msg("Merge failed unexpectedly")
+					logger.Error().Err(err).Msg("Merge failed unexpectedly")
 					continue
 				}
 
 				switch gerr.Response.StatusCode {
 				case http.StatusMethodNotAllowed:
-					logger.Info().Msgf("Merge rejected due to unsatisfied condition %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to unsatisfied condition: %q", gerr.Message)
 					return
 				case http.StatusConflict:
-					logger.Info().Msgf("Merge rejected due to being invalid %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to being invalid: %q", gerr.Message)
 					return
 				default:
-					logger.Error().Err(errors.WithStack(err)).Msgf("Merge failed unexpectedly %q", gerr.Message)
+					logger.Error().Msgf("Merge failed with unexpected status: %d: %q", gerr.Response.StatusCode, gerr.Message)
 					continue
 				}
 			}
 
-			logger.Info().Msgf("Successfully merged pull request for sha %s with message %q", result.GetSHA(), result.GetMessage())
+			logger.Info().Msgf("Successfully merged pull request as SHA %s", sha)
 
-			// Delete ref if owner of BASE and HEAD match
-			// otherwise, its from a fork that we cannot delete
-			if pr.GetBase().GetUser().GetLogin() == pr.GetHead().GetUser().GetLogin() {
+			// if head is qualified (contains ":"), PR is from a fork and we don't have delete permission
```

Comment:
I think we should address this in a later PR, but if the remote fork has bulldozer installed this assumption is no longer correct. I think it was originally here as a workaround from when this operated using oauth tokens. 

We might be able to attempt a delete action, and just fail if it doesn't work. Not blocking, though.

Reply:
If we want to support this, I consider it a new feature that requires more refactoring: even if the fork has Bulldozer installed, we need to look up the installation ID and construct a new client using that ID.

I'm also not sure how useful this would be. It seems odd to need Bulldozer for merges on a repository that mostly submits PRs to an upstream repo, so would the only reason to install it be so that it could delete branches?